src/util.rs

```diff
@@ -24,3 +25,152 @@ pub fn read_int_code_memory(filename: impl AsRef<Path>) -> Vec<i32> {
         .filter_map(Result::ok)
         .collect()
 }
+
+pub struct Permutations {
+    counter: u32,
+    current: Vec<u32>,
+    limit: u32,
+    stack_frames: Vec<IterState>,
+    stack_ptr: u32,
+}
+
+impl Permutations {
+    pub fn new(n: usize) -> Self {
+        let mut arr = vec![0u32; n];
+        let mut stack_frames = Vec::with_capacity(n);
+        let mut limit = 1u32;
+        for i in 0..n {
+            arr[i] = i as u32;
+            limit *= (i as u32) + 1;
+            stack_frames.push(IterState::new(n - i));
+        }
+        Self {
+            counter: 0,
+            current: arr,
+            limit,
+            stack_frames,
+            // Point to the top of the stack
+            stack_ptr: 0,
+        }
+    }
+}
+
+struct IterState {
+    i: isize,
+    depth: usize,
+    in_loop: bool,
+}
+
+impl IterState {
+    fn new(depth: usize) -> Self {
+        Self {
+            depth,
+            i: 0,
+            in_loop: false,
+        }
+    }
+
+    fn copy(&self) -> Self {
+        Self {
+            i: self.i,
+            in_loop: self.in_loop,
+            depth: self.depth,
+        }
+    }
+}
+
+impl Iterator for Permutations {
+    type Item = Vec<u32>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Once we have exceeded the number of possible permutations
+        // return None
+        if self.counter >= self.limit {
+            return None;
+        }
+
+        // At this point the code is pretty much the way it would be. One way to clean this up is
+        // to break up the steps into instructions and map them to numbers or something so when we
+        // go through here we check the stack frame but also the instruction pointer within the frame
+        // that is what instruction are we executing inside the function call.
+        loop {
+            let f = self.stack_frames.get(self.stack_ptr as usize).unwrap();
+            let mut current_frame = f.copy();
+
+            if current_frame.depth == 1 {
+                self.stack_frames[self.stack_ptr as usize] = IterState::new(current_frame.depth);
+                if self.stack_ptr >= 1 {
+                    self.stack_ptr -= 1;
+                }
+                self.counter += 1;
+                return Some(self.current.clone());
+            }
+
+            if !current_frame.in_loop {
+                self.stack_frames[self.stack_ptr as usize] = IterState {
+                    i: current_frame.i,
+                    in_loop: true,
+                    depth: current_frame.depth,
+                };
+                self.stack_ptr += 1;
+                continue;
+            }
+
+            // If we have no idea what i is because we haven't set the loop, we know to start the loop now
+            if current_frame.i >= (current_frame.depth - 1) as isize {
```

Comment:
This comment is outdated. It should say that one we finish the loop, we clean up the current state. In fact, it seems that clearing the current stack frame and reducing the stack pointer is a common thing that could be pulled out into a function